ArduCopter/events.cpp

```diff
@@ -113,10 +113,25 @@ void Copter::failsafe_gcs_check()
     if (should_disarm_on_failsafe()) {
         init_disarm_motors();
     } else {
+        //if in AUTO mode AND failsafe set to "Continue Mission in Auto, then continue mission:
         if (control_mode == AUTO && g.failsafe_gcs == FS_GCS_ENABLED_CONTINUE_MISSION) {
             // continue mission
-        } else if (g.failsafe_gcs != FS_GCS_DISABLED) {
-            set_mode_RTL_or_land_with_pause(MODE_REASON_GCS_FAILSAFE);
+       } 
+        //Engage failsafe if
+        //1. Overriding RC inputs from the GCS.
+        //OR
+        //2. Failsafe set to "always RTL"
+        //OR
+        //3. In GUIDED mode and failsafe set to "RTL only in GUIDED"
+        //OR
+        //4. In either AUTO or GUIDED mode and failsafe set to
+        //  "RTL in AUTO or GUIDED"  
+        else if (failsafe.rc_override_active || 
+            g.failsafe_gcs == FS_GCS_ENABLED_ALWAYS_RTL ||
+            (control_mode == GUIDED && g.failsafe_gcs == FS_GCS_ENABLED_GUIDED) ||
+            ((control_mode == GUIDED || control_mode == AUTO) && g.failsafe_gcs == FS_GCS_ENABLED_GUIDED_OR_AUTO)) {
+                gcs_send_text(MAV_SEVERITY_ERROR,"Lost contact with GCS!");
+                set_mode_RTL_or_land_with_pause(MODE_REASON_GCS_FAILSAFE);
```

Comment:
I'd like this entire else to be something like:

```
switch (g.failsafe_gcs) {
case FS_GCS_ENABLED_CONTINUE_MISSION:
  if (control_mode == AUTO) {
    set_mode_RTL...;
  }
  break;

case FS_GCS_ENABLED_GUIDED:
  if (control_mode != GUIDED) {
    break;
  }

case FS_GCS_ENABLED_GUIDED_OR_AUTO:
  if (control_mode != GUIDED && control_mode != AUTO) {
    break;
  }

case FS_GCS_ENABLED_ALWAYS_RTL: 
  set_mode_RTL...;
  break;
}
```

I think this is much more readable with only repeating the call to set mode once. Regarding the text to the GCS, I agree with it, but it should be above after the *Log_Write_Error* call and I think that it should be a warning, not an error (unless you have found another failsafe where an error is used).